feat(my-account): resubscribe and renew early via modal checkout

miguelpeixe opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Implements the modal checkout on the "Resubscribe" and "Renew now" buttons when viewing a cancelled/expired subscription.

It uses the new `newspackOpenModalCheckout( options )`, proposed in https://github.com/Automattic/newspack-blocks/pull/2121.

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-blocks/pull/2121
2. In a reader account, purchase a subscription product
3. As an admin, cancel the subscription 
4. Let's make sure there are no regressions to the regular flow, so make sure you **don't** have `define( 'NEWSPACK_MY_ACCOUNT_VERSION', '1.0.0' );`
5. As the reader, visit **My Account -> Subscriptions** and navigate to the cancelled subscription
6. Click the Resubscribe button and confirm the regular checkout flow works
7. As an admin, cancel the new subscription
8. Add `define( 'NEWSPACK_MY_ACCOUNT_VERSION', '1.0.0' );` to your config file
9. As the reader, visit the cancelled subscription and click the "Resubscribe" button
10. Confirm the modal renders with the title "Renew subscription"
11. Press escape and confirm the modal closes and the page reloads
12. Click the "Resubscribe" button again and complete the flow
13. Confirm you are redirected to the new subscription after it's done
14. Open your browser console and enter `newspackReaderActivation.getActivities();`
15. Confirm the latest 2 activities are `checkout_completed` with `action_type` as `resubscribe` and `subscription_reactivated` with the `subscription_id` in the payload
16. Click "Renew now" and confirm you can also complete the flow, and the modal is titled "Renew subscription early"
17. Open your browser console and enter `newspackReaderActivation.getActivities();`
18. Confirm the latest 2 activities are `checkout_completed` with `action_type` as `renewal_early` and `subscription_renewal_early` with the `subscription_id` in the payload

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

miguelpeixe committed:
feat(my-account): resubscribe via modal checkout

miguelpeixe committed:
fix: simplify

miguelpeixe committed:
feat: entire flow

miguelpeixe committed:
fix: reload on checkout esc

miguelpeixe committed:
feat: track subscription reactivation

miguelpeixe committed:
fix: set loading state earlier

miguelpeixe committed:
feat: modal checkout for subscription early renewal

miguelpeixe committed:
chore: lint

miguelpeixe committed:
feat: improve modal checkout open options

miguelpeixe committed:
feat: abstract modal checkout button

miguelpeixe committed:
chore: docblock

miguelpeixe committed:
feat: handle subscription renewal id

miguelpeixe committed:
fix: initialize ras array

miguelpeixe committed:
Merge branch 'trunk' into feat/my-account-modal-checkout-resubscribe

miguelpeixe committed:
fix: remove duplicate enqueue

dkoo reviewed (changes requested):
@miguelpeixe this is mostly working well and the flows feel so much smoother now! I tested with a variable subscription product and I think there might be some different handling needed for those. I encountered some unexpected behavior while testing:

1. (Probably not related to the variable product) Getting some warnings when I click the "Resubscribe" button with modal checkout:

```
[22-May-2025 18:05:16 UTC] PHP Warning:  Undefined array key "referer" in /newspack-repos/newspack-blocks/includes/tracking/class-data-events.php on line 163
[22-May-2025 18:05:16 UTC] PHP Deprecated:  str_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated in /newspack-repos/newspack-blocks/includes/tracking/class-data-events.php on line 174
```

2. If you have post-checkout newsletter signup enabled, the signup modal appears after dismissing the success message, but the page redirects anyway. We may want to do something similar to what we did with the [content gate refresh strategy](https://github.com/Automattic/newspack-plugin/blob/feat/my-account-modal-checkout-resubscribe/src/memberships-gate/gate.js#L41-L85), which will only refresh the page once all RAS modals have been dismissed.

3. (Step 13) After the auto-redirect, I was redirected to the page for the parent order, not the subscription itself.

4. (Steps 15 and 16) The subscription ID was not included in the payload for either `checkout_complete` or `subscription_reactivated`/`renewal_early` events. `checkout_complete` had a `variation_id` param, but it was empty.

miguelpeixe committed:
fix: prevent attaching repeated event listeners

miguelpeixe committed:
fix: update attribute name

miguelpeixe committed:
fix: handle overlays on checkout close

miguelpeixe commented:
> (Probably not related to the variable product) Getting some warnings when I click the "Resubscribe" button with modal checkout:

This has been fixed in the `checkoutData` overhaul

> If you have post-checkout newsletter signup enabled, the signup modal appears after dismissing the success message, but the page redirects anyway.

Thank you for catching that! Added overlay handling in 80b088bc578ea53b58db116534dab3cf8ce0fa67

> (Step 13) After the auto-redirect, I was redirected to the page for the parent order, not the subscription itself.
> 
> (Steps 15 and 16) The subscription ID was not included in the payload for either checkout_complete or subscription_reactivated/renewal_early events. checkout_complete had a variation_id param, but it was empty.

These appear to be the same issue. Possibly because I wasn't considering `membership` product type, only `subscription` when producing the checkout data. If the order doesn't include a subscription, the default behavior is to redirect to the order page. Fixed in https://github.com/Automattic/newspack-blocks/pull/2121/commits/33f6b225f536d02592a5b0b2fb57f18e9d8e6fae

dkoo reviewed (approved):
The changes here look good, pending approval of https://github.com/Automattic/newspack-blocks/pull/2121 and conflict resolutions!

miguelpeixe committed:
Merge branch 'trunk' into feat/my-account-modal-checkout-resubscribe
